### PR TITLE
[4.0] Select Modules page

### PR DESCRIPTION
**Pull request by brianteeman**
Inspired by the alternate template proposal which is going nowhere  this changes the new module page from a list to a card view.

~It is massively due some css love and care~

### testing 
dont forget to nmp i for the new css

### before
image

### after
image

### CODE REVIEW

**Comment by Fedik** [comment]
I have two conflicting feelings:
* it looks nicer
* it harder to find needed module

**Comment by HLeithner** [comment]
I have the same feeling as Fedir

**Comment by brianteeman** [comment]
Note: the ordering is the same as before

**Comment by Fedik** [comment]
Couple suggestions for usability improvement:
 - About ordering: it better to make it "row flow", instead of "column flow", then I always know that upper part of alphabet are top, and bottom part of alphabet are bottom. Like now it hard and will be harder with many more modules. 
- Each row and each cell in the row should have the same height, this will prevent "eye jumping" up and down while looking for needed module. It should be "perfect grid" :) 
- Because of needing the same height, the title of cell need to be 1 row (long titles just cut down with overflow), and the description need to reduce in size (2-3 row) and also cut to keep the same height. Perfectly the full description and full title should be expanded "on hover". You know this nice effect in some shops, when you hover on product in grid overview you get expanded cell with product description? :) (something like that https://youtu.be/wHwyDt7bXDY?t=12, sorry I did not found a code example on quick search, technically it just an extra `<div>` with position absolute, that gets bigger on hover)
- Whole cell need to be "clickable" (or better to call it "selectable"), so I quickly can select needed module just by clicking on it without searching for a little "select" button. With this maybe even can remove a "select" button to save some space.

**Comment by brianteeman** [comment]
> Couple suggestions for usability improvement:
> 
> * About ordering: it better to make it "row flow", instead of "column flow", then I always know that upper part of alphabet are top, and bottom part of alphabet are bottom. Like now it hard and will be harder with many more modules.

Agree but I don't know how :(

> * Each row and each cell in the row should have the same height, this will prevent "eye jumping" up and down while looking for needed module. It should be "perfect grid" :)

Change to above should resolve that

> * Because of needing the same height, the title of cell need to be 1 row (long titles just cut down with overflow), and the description need to reduce in size (2-3 row) and also cut to keep the same height. Perfectly the full description and full title should be expanded "on hover". You know this nice effect in some shops, when you hover on product in grid overview you get expanded cell with product description? :) (something like that https://youtu.be/wHwyDt7bXDY?t=12, sorry I did not found a code example on quick search, technically it just an extra `<div>` with position absolute, that gets bigger on hover)

Not sure that will be needed after above

> * Whole cell need to be "clickable" (or better to call it "selectable"), so I quickly can select needed module just by clicking on it without searching for a little "select" button. With this maybe even can remove a "select" button to save some space.

I am **not** a fan of that. It's an accessibility nightmare

**Comment by Fedik** [comment]
It a bit more code than `column-count: 4;` ;)
`column-count` not allow to change "flow". 
Can try [bootstrap grid](https://getbootstrap.com/docs/4.0/layout/grid/) if it enabled in admin template,
or look for use [CSS Flex](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)
```
.card-columns{
  display:flex;
  flex-flow: row;
  flex-wrap: wrap;
}
.card-columns>.card{
  flex: 1 1 250px;
}
```
or [CSS grid](https://css-tricks.com/snippets/css/complete-guide-grid/)
```
.card-columns{
  display: grid;
  grid-template-columns: 1fr, 1fr, 1fr, 1fr;
}
```

> Not sure that will be needed after above

You are right, however I would still limit the description, it will looks more consistent when 1 module has 2 word(row) in description, and another 20 words. But can skip it for now, maybe later someone add.

> I am not a fan of that. It's an accessibility nightmare

I cannot say here much :)
It just wrap a whole card in to `<a>` (maybe with title) and it will be like a list of "buttons", is that still bad? :) 
But this will save really much space, and will make a "select action" faster.

**Comment by brianteeman** [comment]
I switched it to use card-deck etc

The direction is not left to right in rows not columns
Each row is the same height

We are already truncating the module description if it is too long ;)

image

**Comment by Fedik** [comment]
that looks much better :)

> We are already truncating the module description if it is too long ;)

I have missed that :)

**Comment by ghazal** [comment]
I have tested this item :white_check_mark: successfully on f736ad3b6a073ee34b06d35171ee8d87f3123e61<br /><br />Great job. Hope it will be accepted.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27361">issues.joomla.org/tracker/joomla-cms/27361</a>.</sub>

**Comment by TobsBobs** [comment]
I have tested this item :white_check_mark: successfully on f736ad3b6a073ee34b06d35171ee8d87f3123e61<br /><br />Looka nice!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27361">issues.joomla.org/tracker/joomla-cms/27361</a>.</sub>

**Comment by HLeithner** [comment]
Small question shouldn't the 'select' link look more like a button? It looks a bit lost

**Comment by brianteeman** [comment]
@HLeithner thats why I wrote

> It is massively due some css love and care

and why this is an RFC and not a full PR

**Comment by HLeithner** [comment]
> @HLeithner thats why I wrote
> 
> > It is massively due some css love and care
> 
> and why this is an RFC and not a full PR

true, so a button layout would be cool in my opinion ;-)

**Comment by brianteeman** [comment]
Sorry - i should have been clearer - can someone help with that css

~I will give it a try but it's really not my area~

**Comment by Fedik** [comment]
Maybe try add class `btn btn-primary btn-sm` to a button:
screen1

or even drop whole `card-footer` section, and do the button directly after `card-body` as on first example there https://getbootstrap.com/docs/4.0/components/card/ (button class `btn btn-primary`, and remove `mr-2`) :
screen2

or lighter (button class `btn btn-light`) :
screen3

upd: maybe can extend text to more verbose "Select this module", but not important.

**Comment by brianteeman** [comment]
Thanks 
image

**Comment by TobsBobs** [comment]
Should I retest this PR or should I wait because this is a RFC?

**Comment by Fedik** [comment]
hm hm, not really, it not "clickable" ;)

Remove whole section:
```
<div class="card-footer btn btn-primary">
 ...
</div>
```
And add only:
```
<a href="<?php echo Route::_($link); ?>" class="btn btn-primary <?php echo $function ? ' select-link" data-function="' . $this->escape($function) : ''; ?>" aria-label="<?php echo Text::sprintf('COM_MODULES_SELECT_MODULE', $item->name); ?>">
  <?php echo Text::_('JSELECT'); ?>
</a>
```
It will make whole footer as "clickable" button

**Comment by brianteeman** [comment]
Thanks @Fedik 
This is ready for testing now

**Comment by brianteeman** [comment]
I have updated the original post and title

File: administrator/templates/atum/scss/pages/_com_modules.scss [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    grid-template-columns: 1fr 1fr;
+  }
+  @include media-breakpoint-up(md) {
+    grid-template-columns: 1fr 1fr 1fr;
+  }
+  @include media-breakpoint-up(lg) {
+    grid-template-columns: 1fr 1fr 1fr 1fr;
+  }
+  a {
+    color: var(--atum-text-light);
```
Comment (Fedik) [inline administrator/templates/atum/scss/pages/_com_modules.scss]: I think you can remove it now, 
After last changes now it should be correct just with `btn btn-primary` class.

File: administrator/components/com_modules/tmpl/select/default.php [hunk 2]
```diff
@@ -22,19 +22,29 @@
...
-		<small><?php echo $desc; ?></small>
-	</li>
-<?php endforeach; ?>
-</ul>
+				<div class="card-body">
+					<p class="text-muted">
+						<?php echo $desc; ?>
+					</p>
+				</div>
+				<a href="<?php echo Route::_($link); ?>" class="btn btn-primary <?php echo $function ? ' select-link" data-function="' . $this->escape($function) : ''; ?>" aria-label="<?php echo Text::sprintf('COM_MODULES_SELECT_MODULE', $item->name); ?>">
```
Comment (Quy) [inline administrator/components/com_modules/tmpl/select/default.php]:
```suggestion
				<a href="<?php echo Route::_($link); ?>" class="btn btn-primary <?php echo $function ? ' select-link" data-function="' . $this->escape($function) : ''; ?>" aria-label="<?php echo Text::sprintf('COM_MODULES_SELECT_MODULE', $name); ?>">
```

Comment (brianteeman) [reply administrator/components/com_modules/tmpl/select/default.php]: ~thats not correct as $name is not defined and its not consistent with anything else here~

**Comment by Quy** [comment]
Visually I like, however, it requires 2 wheel scrolls to see the entire list and to look for a module, your eyes scan across, diagonal, across, diagonal, and so on vs just up and down.

**Comment by infograf768** [comment]
I agree with @Quy on this one. It is much easier to scan up and down.

**Comment by brianteeman** [comment]
Thats just for scanning titles. When you are also scanning the descriptions it is proven that a shorter line length is more effective.

**Comment by HLeithner** [comment]
In the end I use strg+f to find the module^^

**Comment by Fedik** [comment]
Maybe really try to drop the select button? it will allow to display more rows at once:


can be something like (need to test):
```html
<div class="card mb-4">
<a href="<?php echo Route::_($link); ?>" class="<?php echo $function ? ' select-link" data-function="' . $this->escape($function) : ''; ?>" aria-label="<?php echo Text::sprintf('COM_MODULES_SELECT_MODULE', $item->name); ?>">
  <div class="card-header"> ... </div>
  <div class="card-body"> ... </div> 
</a>
</div>
```
and css
```css
#new-modules-list .card>a{
 text-decoration: none;
 color: inherit;
}
```

I cannot say about an accessibility, 
Maybe can just try to test?

**Comment by brianteeman** [comment]
I am going to stick by the proposal here.

**Comment by Quy** [comment]
Minor styling when in a modal.

add-module

I prefer the current/list format because I can just focus on the module names and look to the right for the description. With the new format, I feel that my eyes are scanning all over the place.

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 169ab46c46f25947a9dc0644fa67c862f78db64c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27361">issues.joomla.org/tracker/joomla-cms/27361</a>.</sub>

**Comment by jwaisner** [comment]
@TobsBobs @ghazal Can you please test the latest again as changes have been made?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27361">issues.joomla.org/tracker/joomla-cms/27361</a>.</sub>

**Comment by brianteeman** [comment]
I really think this is a major ui improvement

**Comment by infograf768** [comment]
I keep with my comment above: this forces to scroll a lot and read from one side of the screen to the other until the specific module is found.
What we have may not be as "beautiful" but it is certainly easier to work with.

**Comment by brianteeman** [comment]
based on comments elsewhere beautiful  is what people want

**Comment by richard67** [comment]
To be honest I also prefer the list style like it is without this PR. But maybe I'm just a bit old fashioned.

**Comment by wilsonge** [comment]
@brianteeman I'd be interested in resurrecting this if you're happy to spend the time making it work with the new redesign?

**Comment by brianteeman** [comment]
I am happy to do it but no one here seemed to like it.

**Comment by brianteeman** [comment]
@wilsonge updated as requested

**Comment by 4digitalworld** [comment]
> 
> 
> ### before
> 
> image


This choice may look much better if there are icons for the items, for example - [Font-Awesome](https://github.com/FortAwesome/Font-Awesome)

**Comment by brianteeman** [comment]
Funny thing is that this is basically the same view from the other template that everyone raved about the design of.

**Comment by 4digitalworld** [comment]
> 
> 
> Funny thing is that this is basically the same view from the other template that everyone raved about the design of.

**I mean, something like this:**

изображение

**Comment by wilsonge** [comment]
Thanks! I think the logical next step is to have screenshots in the boxes - similar to how we do things in the templates view - but this is a good first step and also incorporates the work done by the alternative template team

**Comment by brianteeman** [comment]
Thanks @wilsonge 
Not sure I agree about the screenshots but if someone wants to give a try then fine by me

**Pull request merged**